### PR TITLE
Switch off plan caching for benchmark runs with `--pipeline_metrics`

### LINKED ISSUE #2601: Optimizer runtimes: plotting and metric runs
### Breakdown Plot
In order to analyze the optimizer runtimes (see issues in #2590), the `plot_performance_breakdown.py` script needs to be fixed. Currently, queries obtained from the caches (those do not need to be re-optimized) have an optimization runtime of 0 ns.
When we calculate the average optimization runtime to plot the performance breakdown, we should ignore all zeros as they heavily impact the runtime (especially for queries with few templates, thus high chances of cache hits).

### Metric runs
Adding to the previous issue, we should ensure (warn or fail? Not sure) that people do not use `--metrics` and `--warmup` together as several metrics (e.g., optimizer runtimes) are impacted by pre-filled caches.

### PR DESCRIPTION
**Pull request by dey4ss**
We recently noticed that the performance and optimizer breakdown plots were broken for cached queries: Optimization and translation times were 0 for cached queries. Thus, the plots were completely dominated by the execution runtimes.

After several approaches, it turned out that simply switching off plan caching for metric runs was the best solutions. This is fine since `--metrics` should explicitly measure translation and optimization times. Furthermore, the plotting script was fixed to correctly display the runtimes in ms.

## Example
Selected TPC-DS queries, SF 1, 100 runs (thus, 1 run with translation/optimization, 99 runs are cache hits).
### Status quo
The execution times dominate the plot and the translation/optimization times are invisible. Also, the mean durations are wrong by a factor of 10.
image
### With this PR
image

closes #2601

### CODE REVIEW

File: scripts/plot_performance_breakdown.py [hunk 1]
```diff
@@ -32,7 +34,7 @@
     sum_parse_duration = 0.0
     sum_sql_translation_duration = 0.0
     sum_optimization_duration = 0.0
-    sum_optimizer_rule_durations = {}
+    sum_optimizer_rule_durations = defaultdict(list)
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply scripts/plot_performance_breakdown.py:37]: That's what the defaultdict does, exactly. Just like oerator `[]` for `(unordered_)map`.

File: scripts/plot_performance_breakdown.py [hunk 2]
```diff
@@ -43,17 +45,17 @@
             sum_parse_duration += metrics["parse_duration"]
 
             for statement in metrics["statements"]:
+                if statement["query_plan_cache_hit"]:
+                    exit("When generating metrics, plan caching should be switched off.")
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply scripts/plot_performance_breakdown.py:49]: Not anymore with this PR. However, I thought a sanity check does not hurt and ensures no plots are generated with old JSONs. Can turn this to an `assert` though if you prefer.

File: src/benchmarklib/benchmark_runner.cpp [hunk 3]
```diff
@@ -545,7 +553,7 @@ cxxopts::Options BenchmarkRunner::get_basic_cli_options(const std::string& bench
     ("visualize", "Create a visualization image of one LQP and PQP for each query, do not properly run the benchmark", cxxopts::value<bool>()->default_value("false"))  // NOLINT(whitespace/line_length)
     ("verify", "Verify each query by comparing it with the SQLite result", cxxopts::value<bool>()->default_value("false"))  // NOLINT(whitespace/line_length)
     ("dont_cache_binary_tables", "Do not cache tables as binary files for faster loading on subsequent runs", cxxopts::value<bool>()->default_value("false"))  // NOLINT(whitespace/line_length)
-    ("metrics", "Track more metrics (steps in SQL pipeline, system utilization, etc.) and add them to the output JSON (see -o)", cxxopts::value<bool>()->default_value("false"))  // NOLINT(whitespace/line_length)
+    ("metrics", "Track more metrics (steps in SQL pipeline, system utilization, etc.) and add them to the output JSON (see -o). Tracking metrics switches off plan caching.", cxxopts::value<bool>()->default_value("false"))  // NOLINT(whitespace/line_length)
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/benchmarklib/benchmark_runner.cpp:556]: Gonna have a look at it

File: scripts/plot_performance_breakdown.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dey4ss) [reply scripts/plot_performance_breakdown.py:37]: 🐍

File: src/benchmarklib/benchmark_runner.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/benchmarklib/benchmark_runner.cpp:556]: Seems like `--metrics` tracks (i) SQL pipeline metrics and optimizer rule durations, (ii) segment statistics/access counters from `meta_segments` (ii) system usage information from `meta_systems_utilization` (see below).

https://github.com/hyrise/hyrise/blob/a23585c4360efda5413d7dbce86b7b6f2292ed3d/src/lib/utils/meta_tables/meta_system_utilization_table.cpp#L19-L29

Turning caching off should not be too bad I guess (we should only lose the size of the caches in the memory usage). However, we could split `--metrics` to `--pipeline_metrics` ((i), what we plot with `plot_performance_breakdown.py`) and `--system_metrics` ((ii) segment statistics, (iii) usage info) if you prefer. Thus, no caching only for the first option.

[the remark this answers is not in the document]
Comment (Bouncner) [reply src/benchmarklib/benchmark_runner.cpp:556]: The two flags sound like the best solution to me.

File: scripts/plot_performance_breakdown.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bouncner) [reply scripts/plot_performance_breakdown.py]: Assert or exit both work for me. But maybe add a comment that this SHOULD never happen?

File: src/benchmarklib/benchmark_runner.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/benchmarklib/benchmark_runner.cpp:556]: Only the pipeline metrics are dumped in the "metrics" key for each benchmark statement run in the result JSON. So I'm also gonna rename it to "pipeline_metrics".

[the remark this answers is not in the document]
Comment (Bouncner) [reply src/benchmarklib/benchmark_runner.cpp]: Okay.

**Review by Bouncner — APPROVED** [review]
LGTM, but when search for `metrics` there was a ton of hits in the code base. Did you have a quick view if certain mentions/usages of `metrics` should be clarified now that we have two different metrics?

**Comment by dey4ss** [comment]
We have two `metrics` flags now, but we have already had different metrics objects earlier (`SQLPipelineMetrics` – the one mostly referenced by the term "metrics" –, `OptimizerRuleMetrics` – which are a member of the first –, `TableGenerationMetrics`, ...). So due to the object type and context it should be clear in most cases (pipeline metrics for SQLPipeline, SQL execution, benchmark items, etc., optimizer rule metrics for optimizer, table generation metrics for table generator). At first glance, I did not spot an obvious example where it was unclear what metrics related to.

**Pull request merged**